### ISSUE TITLE
Proposal to Add PPO + Mamba SSM to CleanRL

### ISSUE THREAD

**Issue by Tornadosky** [vwxyzjn/cleanrl#510]
# Proposal to Add PPO + Mamba SSM to CleanRL

## Issue Type
- [ ] Bug Report
- [x] Feature Request / Algorithm Implementation
- [ ] Documentation
- [ ] Question

## Overview
This issue proposes adding a new algorithm implementation: PPO integrated with the [Mamba State-Space Model (SSM)](https://arxiv.org/abs/2312.00752). Mamba is a recent neural network architecture designed for efficient sequence modeling, demonstrating significant improvements in computational speed and memory usage over traditional recurrent models (LSTM, GRU) and Transformer-based models such as the existing [ppo_trxl](https://docs.cleanrl.dev/rl-algorithms/ppo-trxl/) implementation.

## Motivation
- Faster training speeds
- Lower GPU memory usage compared to Transformer-XL
- Good performance in POMDP and memory-based tasks

## Environments Tested
- ProofOfMemory Environments
- MiniGrid (Memory, DoorKey)
- Classical Control (masked): LunarLander, CartPole
- MuJoCo continuous control tasks: HalfCheetah, Walker2d, Hopper
- Atari games: Breakout, Pong

## Current Behavior
- No current implementation of PPO + Mamba SSM in CleanRL

## Expected Behavior
- Implementation of PPO integrated with Mamba SSM
- Performance benchmarks showing training efficiency and final performance
- Documentation on usage and hyperparameter tuning

## Planned Contribution Steps
1. Complete comprehensive benchmarks following CleanRL's contribution guidelines
2. Utilize the RLops utility for rigorous performance and regression analysis
3. Update the CleanRL documentation to clearly present benchmark results and learning curves
4. Submit PR with implementation and documentation

I already have experiments in these environments and also the comparison of efficiency metrics. I tested both [Mamba](https://arxiv.org/abs/2312.00752) and [Mamba-2](https://arxiv.org/abs/2405.21060) models.

@vwxyzjn , I plan to follow the way [ppo_trxl](https://docs.cleanrl.dev/rl-algorithms/ppo-trxl/) was integrated. What do you think?

**Comment by BertrandDecoster** [comment]
I support this. A mamba implementation would be very beneficial

**Comment by MarcoMeter** [comment]
Seeing some scores on Endless Memory Gym would be amazing.

**Comment by gspeter-max** [comment]
## Proposal to Implement PPO + Mamba in CleanRL

I would like to work on implementing PPO + Mamba (both Mamba-1 and Mamba-2) for CleanRL. Based on the discussion in this issue, I've outlined a design approach and would love feedback from the maintainers before proceeding.

### Design Overview

**Goal:** Add PPO + Mamba as a new algorithm variant, following CleanRL's single-file philosophy while supporting multiple environment types.

**Proposed Implementation:**

#### File Structure
```
cleanrl/
├── ppo_atari_mamba.py          # Atari (Breakout, Pong, etc.)
├── ppo_continuous_mamba.py     # MuJoCo (HalfCheetah, Walker2d, etc.)
├── ppo_memory_mamba.py         # Memory tasks (MiniGrid, Memory Gym)

cleanrl_utils/
├── mamba_agent.py              # Shared Mamba Agent class
├── mamba_config.py             # Mamba hyperparameters

requirements/
└── requirements-mamba.txt      # mamba-ssm dependency

docs/rl-algorithms/
└── ppo-mamba.md                # Documentation
```

#### Key Design Decisions

1. **Modular files approach** - Separate files for each environment type, sharing a common Mamba agent utility
2. **Mamba library** - Use official `mamba-ssm` package (not from-scratch implementation)
3. **Both Mamba variants** - Support Mamba-1 and Mamba-2 via hyperparameter selection
4. **Full environment support** - Atari + MuJoCo + memory environments (MiniGrid, Memory Gym)

#### Mamba Agent Design

The shared `MambaAgent` class would:
- Replace LSTM's (hidden, cell) state with single unified state
- Use CNN encoder for Atari, MLP for MuJoCo/memory
- Maintain compatible interface with existing PPO training loops

#### Comparison Points

I plan to benchmark against:
- `ppo_atari_lstm.py` - LSTM baseline
- `ppo_trxl/ppo_trxl.py` - Transformer-XL baseline
- Focus on: training speed, memory usage, final performance

### Questions for Maintainers

1. **Is the modular files approach acceptable?** CleanRL typically uses single files, but the Mamba agent would be duplicated across Atari/MuJoCo/memory variants.

2. **Should this be in a subdirectory like `ppo_mamba/`** (similar to `ppo_trxl/`) rather than top-level files?

3. **Are there specific environments** from the issue author's experiments that should be prioritized?

4. **Any concerns about adding `mamba-ssm` as a dependency?** It's PyPI-packaged and actively maintained.

### Next Steps

If this approach is approved:
1. Create detailed implementation plan
2. Implement `cleanrl_utils/mamba_agent.py` first
3. Build `ppo_atari_mamba.py` as proof-of-concept
4. Run RLops benchmarks
5. Submit PR with full documentation

Looking forward to feedback! @vwxyzjn @Tornadosky
